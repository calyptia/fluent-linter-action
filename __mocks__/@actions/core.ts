export const setFailed = jest.fn();
export const getInput = jest.fn();
export const setOutput = jest.fn();
export const debug = jest.fn();
export const error = jest.fn();
