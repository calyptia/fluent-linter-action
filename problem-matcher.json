{
  "problemMatcher": [
    {
      "owner": "fluent-linter",
      "pattern": [
        {
          "regexp": "(.*\\/[^\\/]*):\\s(\\d+):(\\d+)\\s+(error|warning|info)\\s+(\\w+)\\s+(.*)$",
          "file": 1,
          "line": 2,
          "column": 3,
          "severity": 4,
          "message": 6
        }
      ]
    }
  ]
}
