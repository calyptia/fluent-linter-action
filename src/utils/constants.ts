import { getBorderCharacters } from 'table';

export const SUCCESS_SYMBOL = 'ğŸ’«';
export const SHRUG_SYMBOL = 'Â¯\\_(ãƒ„)_/Â¯';
export const CALYPTIA_API_ENDPOINT = 'https://cloud-api.calyptia.com';

export const CALYPTIA_API_VALIDATION_PATH = 'v1/config_validate/fluentbit';

export const PROBLEM_MATCHER_FILE_LOCATION = '.github/problem-matcher.json';

export const NO_STYLES_IN_TABLE = {
  border: getBorderCharacters('void'),
  columnDefault: {
    paddingLeft: 0,
    paddingRight: 1,
  },
  drawHorizontalLine: (): boolean => false,
};
