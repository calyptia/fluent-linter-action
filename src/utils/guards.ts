import type { FullError } from './normalizeErrors';

export const isString = (value: unknown): value is string => typeof value === 'string';

export const isFullError = (content: unknown): content is FullError => Array.isArray(content) && content.length === 3;
