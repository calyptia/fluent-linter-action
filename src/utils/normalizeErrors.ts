import { join } from 'path';
import type { ValidatedConfigV2 } from '../../generated/calyptia';

export type IdError = [id: string, message: string];

export type FullError = [line: number, col: number, reason: string];
type ErrorGroup = IdError | FullError;
export type Annotation = { filePath: string; errors: ErrorGroup[] };

declare let process: {
  env: {
    GITHUB_WORKSPACE: string;
  };
};
export function getRelativeFilePath(filePath: string): string {
  const relativePath = filePath.replace(join(process.env.GITHUB_WORKSPACE, '/'), '');

  // We add the ./ to the relative path to match the problem matcher.
  return filePath.startsWith('/') ? relativePath : join('./', relativePath);
}
export function normalizeErrors(filePath: string, errors: ValidatedConfigV2['errors']): Annotation[] {
  const annotations = [] as Annotation[];

  const relativeFilePath = getRelativeFilePath(filePath);

  const { runtime, ...rest } = errors;

  if (runtime.length) {
    for (const error of runtime) {
      annotations.push({ filePath: relativeFilePath, errors: error.errors.map((err: string) => [error.id, err]) });
    }
  }
  for (const command in rest) {
    const issues = rest[command as keyof typeof rest];

    if (Object.keys(issues).length) {
      const errors = Object.entries(issues).reduce((memo, [, errs]) => {
        return [...memo, ...errs.map(({ id, errors }) => [id, errors.join('\n')] as ErrorGroup)];
      }, [] as ErrorGroup[]);

      annotations.push({ filePath: relativeFilePath, errors });
    }
  }

  return annotations;
}
