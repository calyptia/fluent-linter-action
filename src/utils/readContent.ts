import fs from 'fs';
import { promisify } from 'util';

const readFile = promisify(fs.readFile);
export const readContent = async (filePath: string): Promise<string> => {
  const content = await readFile(filePath, { encoding: 'utf-8' });
  return content;
};
