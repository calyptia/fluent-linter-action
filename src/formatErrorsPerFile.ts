import type { Annotation } from './utils/normalizeErrors';
import { table } from 'table';
import { NO_STYLES_IN_TABLE } from './utils/constants';
import { isFullError } from './utils/guards';
import type { FluentBitSchema } from '@calyptia/fluent-bit-config-parser';

const DEFAULT_LINE_AND_COLUMN = '0:0';
const ISSUE_LEVEL = 'error';
export function formatErrorsPerFile(
  filePath: string,
  errorGroups: Annotation['errors'],
  schema?: FluentBitSchema
): string {
  const data = [] as string[][];

  for (const error of errorGroups) {
    let content = [];
    console.log('error:', error);
    if (isFullError(error)) {
      const [line, col, message] = error;
      content = [`${filePath}:`, `${line}:${col}`, ISSUE_LEVEL, 'LINTER', message];
    } else {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const tokens = schema!.getTokensBySectionId(error[0]);

      if (tokens) {
        content = [`${filePath}:`, `${tokens[0].line}:${tokens[0].col}`, ISSUE_LEVEL, 'LINTER', error[1]];
      } else {
        content = [`${filePath}:`, DEFAULT_LINE_AND_COLUMN, ISSUE_LEVEL, 'LINTER', error[1]];
      }
    }
    data.push(content);
  }
  return table(data, NO_STYLES_IN_TABLE);
}
