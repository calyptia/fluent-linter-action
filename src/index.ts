import { getInput, setFailed, debug } from '@actions/core';
import * as glob from '@actions/glob';
import { readContent } from './utils/readContent';
import { FluentBitSchema, TokenError } from '@calyptia/fluent-bit-config-parser';
import fetch from 'node-fetch';
import { CALYPTIA_API_ENDPOINT, CALYPTIA_API_VALIDATION_PATH, PROBLEM_MATCHER_FILE_NAME } from './utils/constants';
import { Annotation, FieldErrors, normalizeErrors, relativeFilePath } from './utils/normalizeErrors';
import { formatErrorsPerFile } from './formatErrorsPerFile';
import { resolve } from 'path';
export enum InputValues {
  CONFIG_LOCATION_GLOB = 'CONFIG_LOCATION_GLOB',
  CALYPTIA_API_KEY = 'CALYPTIA_API_KEY',
}

type ValidationResponse = { errors?: FieldErrors };

const getActionInput = () => {
  return Object.keys(InputValues).reduce((memo, prop) => {
    const value = getInput(prop);
    return { ...memo, [prop]: value };
  }, {} as Record<keyof typeof InputValues, string>);
};

export const main = async (): Promise<void> => {
  const input = getActionInput();

  const globber = await glob.create(input.CONFIG_LOCATION_GLOB, { matchDirectories: false });

  let annotations = [] as Annotation[];
  const location = resolve(__dirname, PROBLEM_MATCHER_FILE_NAME);
  console.log(`::add-matcher::${location}`);

  for await (const filePath of globber.globGenerator()) {
    debug(`evaluating file ${filePath}`);

    const content = await readContent(filePath);

    if (FluentBitSchema.isFluentBitConfiguration(content)) {
      debug(`File ${filePath} seems to be Fluent Bit config, validating...`);

      const URL = `${CALYPTIA_API_ENDPOINT}/${CALYPTIA_API_VALIDATION_PATH}`;

      const headers = {
        'Content-Type': 'application/json',
        'x-project-token': input.CALYPTIA_API_KEY,
      };

      try {
        const config = new FluentBitSchema(content, filePath);
        const response = (await fetch(URL, {
          method: 'POST',
          body: JSON.stringify(config.schema),
          headers,
        })) as Response;

        const data = (await response.json()) as unknown as ValidationResponse;

        if (response.status === 200) {
          debug(`[${filePath}]: ${JSON.stringify(data)}`);

          if (data.errors) {
            const errors = normalizeErrors(filePath, data.errors);

            debug(`${filePath}, Found errors: ${JSON.stringify(errors, null, 2)}`);
            annotations = [...annotations, ...errors];
          }
        } else {
          setFailed(`The request failed:  status: ${response.status}, data: ${JSON.stringify(data)}`);
        }
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } catch (e) {
        if (e instanceof TokenError) {
          const { filePath: _filePath, line, col, message } = e as TokenError;
          const errorReport = formatErrorsPerFile(relativeFilePath(_filePath), [['PARSE', [[line, col, message]]]]);
          console.log(errorReport);
        } else {
          setFailed((e as Error).message);
        }
        setFailed('We found an error, please check, please check your logs');
      }
    }
  }

  if (annotations.length) {
    const groupedByFile = annotations.reduce((memo, { filePath, errorGroups }) => {
      memo[filePath] = memo[filePath] ? [...memo[filePath], ...errorGroups] : errorGroups;

      return memo;
    }, {} as Record<string, unknown[]>);
    for (const file in groupedByFile) {
      console.log(formatErrorsPerFile(file, groupedByFile[file] as Annotation['errorGroups']));
    }
    setFailed('We found errors in your configurations. Please check your logs');
  }
};
