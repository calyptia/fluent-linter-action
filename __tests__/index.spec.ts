import { main, InputValues } from '../src';
import { setFailed, getInput, setOutput } from '../__mocks__/@actions/core';
import nock from 'nock';
import failCase from '../__fixtures__/scenarios/failed_case.json';
import { CALYPTIA_API_ENDPOINT, CALYPTIA_API_VALIDATION_PATH } from '../src/utils/constants';
import { mockConsole, unMockConsole } from './helpers';
import { problemMatcher } from '../.github/problem-matcher.json';
describe('fluent-linter-action', () => {
  let consoleLogMock: jest.Mock;
  const mockedInput = {
    [InputValues.GITHUB_TOKEN]: 'GITHUB_TOKEN',
    [InputValues.CALYPTIA_API_KEY]: 'API_TOKEN',
    [InputValues.CONFIG_LOCATION_GLOB]: '__fixtures__/*.conf',
  };

  process.env.GITHUB_WORKSPACE = __dirname;
  beforeAll(() => {
    getInput.mockImplementation((key: Partial<InputValues>) => {
      return mockedInput[key];
    });

    consoleLogMock = mockConsole('log');
  });

  afterAll(() => {
    unMockConsole('log');
  });

  afterEach(() => {
    getInput.mockClear();
    setFailed.mockClear();
    setOutput.mockClear();
    consoleLogMock.mockClear();
  });

  it('runs when provided the correct input: 1 correct, 1 with errors, the last one fails with server error', async () => {
    const client = nock(CALYPTIA_API_ENDPOINT)
      ['post']('/' + CALYPTIA_API_VALIDATION_PATH)
      .reply(200, JSON.stringify({ config: {} }));
    client['post']('/' + CALYPTIA_API_VALIDATION_PATH).reply(200, failCase);

    client['post']('/' + CALYPTIA_API_VALIDATION_PATH).replyWithError(new Error('Server Error'));

    await main();
    // expect(setFailed).toHaveBeenCalled();
    expect(setFailed.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "something went very wrong \\"request to https://cloud-api.calyptia.com/v1/config_validate/fluentbit failed, reason: Server Error\\"",
        ],
        Array [
          "We found errors in your configurations. Please check your logs",
        ],
      ]
    `);
    expect(consoleLogMock.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "::add-matcher::.github/problem-matcher.json",
        ],
        Array [
          "<PROJECT_ROOT>/__fixtures__/invalid.conf: 0:0 error john   cannot initialize input plugin: john 
      <PROJECT_ROOT>/__fixtures__/invalid.conf: 0:0 error syslog Unknown syslog mode abc              
      <PROJECT_ROOT>/__fixtures__/invalid.conf: 0:0 error parser missing 'key_name'                   
      ",
        ],
      ]
    `);
    const [
      {
        pattern: [{ regexp }],
      },
    ] = problemMatcher;

    const [issues] = consoleLogMock.mock.calls[1] as string[];

    const errors = issues.split('\n');

    errors.pop(); // We end up with a last line jump for format that we don't want in the loop.

    for (const error of errors) {
      const issue = error.match(new RegExp(regexp));

      if (issue) {
        const [, file, line, column, severity, , message] = issue;

        expect({
          file,
          line,
          column,
          severity,
          message,
        }).toMatchSnapshot(`${error}`);
      }
    }

    expect(client.isDone()).toBe(true);
  });
});
