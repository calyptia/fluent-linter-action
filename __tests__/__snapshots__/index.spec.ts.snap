// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fluent-linter-action runs when provided the correct input: 1 correct, 1 with errors, the last one fails with server error: /Users/gago/base/fluent-linter-action/__fixtures__/invalid.conf: 0:0 error john   cannot initialize input plugin: john  1`] = `
Object {
  "column": "0",
  "file": "<PROJECT_ROOT>/__fixtures__/invalid.conf",
  "line": "0",
  "message": "cannot initialize input plugin: john ",
  "severity": "error",
}
`;

exports[`fluent-linter-action runs when provided the correct input: 1 correct, 1 with errors, the last one fails with server error: /Users/gago/base/fluent-linter-action/__fixtures__/invalid.conf: 0:0 error parser missing 'key_name'                    1`] = `
Object {
  "column": "0",
  "file": "<PROJECT_ROOT>/__fixtures__/invalid.conf",
  "line": "0",
  "message": "missing 'key_name'                   ",
  "severity": "error",
}
`;

exports[`fluent-linter-action runs when provided the correct input: 1 correct, 1 with errors, the last one fails with server error: /Users/gago/base/fluent-linter-action/__fixtures__/invalid.conf: 0:0 error syslog Unknown syslog mode abc               1`] = `
Object {
  "column": "0",
  "file": "<PROJECT_ROOT>/__fixtures__/invalid.conf",
  "line": "0",
  "message": "Unknown syslog mode abc              ",
  "severity": "error",
}
`;
